lib/zlib.toit

```diff
@@ -256,72 +292,136 @@ class ZlibBackend_ implements Backend_:
 
   static STATE-READY-TO-READ_  ::= 1 << 0
   static STATE-READY-TO-WRITE_ ::= 1 << 1
 
   constructor .backend_:
-    reader = CompressionReader.private_
-    reader.wrapped_ = this
     add-finalizer this::
       this.uninit_
 
   /**
   A reader that can be used to read the compressed or decompressed data output
     by the Encoder or Decoder.
+
+  By default the $CompressionReader blocks until there is data available.
+  Use $CompressionReader.read with the '--wait' flag set to false to get an empty
+    ByteArray when the buffers are empty, and a call to the write method is
+    needed.
   */
-  reader/CompressionReader
+  in -> CompressionReader:
+    if not in_: in_ = CompressionReader.private_ this
+    return in_
+
+  /**
+  Deprecated. Use $in instead.
+  */
+  reader -> CompressionReader:
+    return in
+
+  /**
+  A writer that can be used to write data to the Encoder or Decoder.
+
+  If the writers $io.Writer.try-write returns 0, then that means that the
+    read method must be called because the buffers are full.
+  */
+  out -> io.CloseableWriter:
+    if not out_: out_ = CompressionWriter_.private_ this
+    return out_
 
   read_ --wait/bool -> ByteArray?:
-    if closed-read_: return null
+    return in.read --wait=wait
+
+  consume_ -> ByteArray?:
+    if buffered_:
+      result := buffered_
+      buffered_ = null
+      return result
     result := backend_.read
-    while result and wait and result.size == 0:
-      state_ &= ~STATE-READY-TO-READ_
-      state_ |= STATE-READY-TO-WRITE_
-      signal_.raise
-      signal_.wait: state_ & STATE-READY-TO-READ_ != 0
+    while result and result.size == 0:
+      wait-for-more-data_
       result = backend_.read
     return result
 
-  close-read_ -> none:
-    if not closed-read_:
-      closed-read_ = true
-      if closed-write_:
-        uninit_
-      state_ |= STATE-READY-TO-WRITE_
-      signal_.raise
+  /**
+  Whether the backend needs more input.
+
+  This is done by asking the backend and then, potentially, storing the
+    returned data in this instance.
+  */
+  backend-needs-input_ -> bool:
+    assert: not buffered_
+    data := backend_.read
+    if not data:
+      // The backend is closed.
+      return false
+    if data.size != 0:
+      buffered_ = data
+      // Data is available.
+      return false
+    // No data. The backend needs input.
+    return true
+
+  close-reader_ -> none:
+    if out.is-closed:
+      uninit_
+    state_ |= STATE-READY-TO-WRITE_
+    signal_.raise
 
   /**
   Writes data to the compressor or decompressor.
   If $wait is false, it may return before all data has been written.
     If it returns zero, then that means the read method must be called
     because the buffers are full.
+
+  Deprecated. Use out.try-write or out.write instead.
   */
   write --wait/bool=true data/io.Data from/int=0 to/int=data.byte-size -> int:
-    if closed-read_: throw "READER_CLOSED"
+    if not wait: return try-write_ data from to
```

Comment:
This function no longer throws if the read side was closed?

Reply:
The check is in the `try-write_`.